packages/deployer/src/compiler.ts

```diff
@@ -109,9 +126,14 @@ export class Compiler {
             shouldCompile = true;
         } else {
             const currentArtifact = currentArtifactIfExists as ContractArtifact;
-            shouldCompile =
-                currentArtifact.networks[this._networkId].optimizer_enabled !== this._optimizerEnabled ||
-                currentArtifact.networks[this._networkId].source_tree_hash !== sourceTreeHashHex;
+            if (_.isUndefined(currentArtifact.versions) || _.isUndefined(currentArtifact.versions[this._versionName])) {
+                shouldCompile = true;
+            } else {
+                shouldCompile =
+                    currentArtifact.schemaVersion !== '2.0.0' ||
+                    !_.isEqual(currentArtifact.versions[this._versionName].compiler.settings, this._compilerSettings) ||
+                    currentArtifact.versions[this._versionName].sourceTreeHashHex !== sourceTreeHashHex;
```

Comment:
Let's assign `currentArtifact.versions[this._versionName]` to a variable called `versionArtifact` above instead of indexing in everywhere. That will help with code readability.